src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -41,20 +46,166 @@ module Database.LSMTree.Internal.Run.Construction (
   ) where
 
 import           Control.Exception (assert)
+import           Control.Monad (forM_, unless)
+import           Control.Monad.ST.Strict
 import           Data.Bits (Bits (..))
 import qualified Data.ByteString.Builder as BB
 import           Data.ByteString.Short.Internal (ShortByteString (SBS))
 import           Data.Foldable (Foldable (..))
 import           Data.Maybe (fromMaybe)
 import           Data.Monoid (Dual (..))
 import           Data.Primitive.ByteArray (ByteArray (..), sizeofByteArray)
+import           Data.STRef
 import           Data.Word (Word16, Word32, Word64, Word8)
 import           Database.LSMTree.Internal.Entry (Entry (..), onBlobRef,
                      onValue)
+import           Database.LSMTree.Internal.Run.BloomFilter (Bloom, MBloom)
+import qualified Database.LSMTree.Internal.Run.BloomFilter as Bloom
+import           Database.LSMTree.Internal.Run.Index.Compact (CompactIndex,
+                     MCompactIndex)
+import qualified Database.LSMTree.Internal.Run.Index.Compact as Index
 import           Database.LSMTree.Internal.Serialise (SerialisedKey,
                      serialisedKey, shortByteStringFromTo, sizeofKey16,
                      sizeofKey64, topBits16)
 
+{-------------------------------------------------------------------------------
+  Incremental, in-memory run construction
+-------------------------------------------------------------------------------}
+
+-- | A mutable structure that accumulates k\/op pairs and yields pages.
+--
+-- Use 'new' to start run construction, add new key\/operation pairs to the run
+-- by using 'addFullKOp' and co, and complete run construction using
+-- 'unsafeFinalise'.
+data MRun s = MRun {
+      mbloom               :: !(MBloom s SerialisedKey)
+    , mindex               :: !(MCompactIndex s)
+    , indexChunksRef       :: !(STRef s [[Index.Chunk]])
+    , currentPageRef       :: !(STRef s PageAcc)
+    , rangeFinderPrecision :: !Int
+    }
+
+-- | @'new' npages@ starts an incremental run construction.
+--
+-- @nentries@ and @npages@ should be an upper bound on the expected number of
+-- entries and pages in the output run.
+new :: Int -> Int -> ST s (MRun s)
+new nentries npages = do
+    mbloom <- Bloom.newEasy 0.1 nentries -- TODO(optimise): tune bloom filter
```

Comment:
Currently defaults to using `suggestSizing` for an FPR of `0.1`. Does that seem reasonable @dcoutts ?

Reply:
It's fine for now. We can adjust it, and we will switch to Monkey size calculations later.